sympy/tensor/tensor.py

```diff
@@ -3988,6 +3999,111 @@ def __iter__(self):
                 raise ValueError("No iteration on abstract tensors")
             return self.data.__iter__()
 
+    def _dedupe_indices(self, new, exclude):
+        """
+        exclude: set
+        new: TensExpr
+
+        If ``new`` has any dummy indices that are in ``exclude``, return a version
+        of new with those indices replaced. If no replacements are needed,
+        return None
+        """
+        inds_self = set(exclude)
+        dums_new = set(get_dummy_indices(new))
+
+        conflicts = dums_new.intersection(inds_self)
+        if len(conflicts) == 0:
+            return None
+
+        """
+        ``self_args_free`` is to be passed to ``_IndexStructure._get_generator_for_dummy_indices()``.
+        Since the latter does not use the index position for anything, we just
+        set it as ``None`` here.
+        """
+        inds_self.update(dums_new)
+        self_args_free = [(i, None) for i in inds_self]
+        gen = self._index_structure._get_generator_for_dummy_indices(self_args_free)
+        repl = {}
+        for d in conflicts:
+            if -d in repl.keys():
+                continue
+            newname = gen(d.tensor_index_type)
+            new_d = d.func(newname, *d.args[1:])
+            repl[d] = new_d
+            repl[-d] = -new_d
+
+        if len(repl) == 0:
+            return None
+
+        new_renamed = new._replace_indices(repl)
+        return new_renamed
+
+    def _dedupe_indices_in_rule(self, rule):
+        """
+        rule: dict
+
+        This applies self._dedupe_indices on all values of rule.
+        """
+        index_rules = {k:v for k,v in rule.items() if isinstance(k, TensorIndex)}
+        other_rules = {k:v for k,v in rule.items() if k not in index_rules.keys()}
+        exclude = set(self.get_indices())
+
+        newrule = {}
+        newrule.update(index_rules)
+        exclude.update(index_rules.keys())
+        exclude.update(index_rules.values())
+        for old, new in other_rules.items():
+            new_renamed = self._dedupe_indices(new, exclude)
+            if old == new or new_renamed is None:
+                newrule[old] = new
+            else:
+                newrule[old] = new_renamed
+                exclude.update(get_indices(new_renamed))
+        return newrule
+
+    def _eval_subs(self, old, new):
+        """
+        This handles the fact that if a dummy index in new is the same as an
+        index in self (which is not present in old), the dummy index in new
+        must be renamed.
+        """
+
+        if not isinstance(new, TensExpr):
+            return None
+
+        new_renamed = self._dedupe_indices(new, self.get_indices())
+        if new_renamed is None:
+            return None
+        else:
+            return self.subs({old: new_renamed})
+
+    def _xreplace(self, rule):
```

Comment:
It shouldn't be necessary to redefine the `_xreplace` method. This method is supposed to be inherited from Basic and behave the same for all subclasses.

Reply:
The reason I wanted to fix `xreplace` is that `match` internally uses `xreplace`, and `match` is in turn used by `replace`. Would you prefer instead redefining `match` for `TensMul`? I guess that is also doable.

I initially (not in this PR) tried some index deduplication tricks in `TensMul.__new__`, but what I ended up with seemed quite complicated.